Support SameSite

benmccallum opened the pull request:
Adds support for parsing the SameSite value out of the header properly too.
* Requires netfx472 (which has latest enum value for `SameSiteMode.None`)

benmccallum commented:
Would you be able to merge this and get out a new NuGet version @AndersAbel? I'd prefer not to have to release my own pkg and cause fragmentation.

Will drop back the Owin deps changes as that's probably not needed.

benmccallum commented:
One thing I did notice though that was weird is, one of the tests fails. I confirmed it fails on your current `master` branch too. I'm not sure what it's for, but it's essentially that the `FromHeader` prop on `HttpCookie` doesn't seem to exist anymore. Doesn't in 472 or 452 though, so 🤷‍♂

benmccallum commented:
Potentially I should just remove that code, as it'd never be doing anything. Let me know.

benmccallum commented:
UPDATE: `UseOpenIdConnect` fix requires latest Owin 4.1 updates according to MS blog. Not sure if that belong here or not. Probably not.

AndersAbel commented:
I was living in the illusion that I could leave this package as it is. Obviously not, thanks to SameSite.

The recommended mitigation for the "cookie monster" is to use a SystemWebCookieManager that connects the Owin cookie API directly to the System.Web cookie API. But I guess there are people that still want their Owin applications to no be tied to System.Web. @benmccallum what's the background that you still use this package?

I think updating the dependencies to 4.1.0 (to force people to use a supported version of Katana) makes sense. And obviously to release a new version. What about writing in the release notes that the SystemWebCookieManager is a better way to solve this?

// @tratcher any thoughts on how to help people migrate off this package and to SystemWebCookieManager instead?

benmccallum commented:
Hey @AndersAbel, thanks for your reply. 

Our situation is that we are still supporting an MVC5 app that uses Auth0 and `UseOpenIdConnectAuthentication`. Our config is:

```
// Enable the Cookie saver middleware to work around a bug in the OWIN implementation
app.UseKentorOwinCookieSaver();

// Set Cookies as default authentication type
app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);
app.UseCookieAuthentication(new CookieAuthenticationOptions
{
    AuthenticationType = CookieAuthenticationDefaults.AuthenticationType,
    LoginPath = new PathString("/login"),
    CookieSecure = CookieSecureOption.SameAsRequest,
    SlidingExpiration = true,
    CookieManager = new SameSiteCookieManager(new SystemWebCookieManager()) // I've added this!!
});

// Configure Auth0 authentication
app.UseOpenIdConnectAuthentication(new OpenIdConnectAuthenticationOptions
{
    AuthenticationType = "Auth0",
    Authority = $"https://{auth0Domain}",
    ClientId = auth0ClientId,
    ClientSecret = auth0ClientSecret,
    RedirectUri = auth0OwinRedirectUrl,
    PostLogoutRedirectUri = auth0OwinLogoutRedirectUrl,                
    ResponseType = OpenIdConnectResponseType.CodeIdTokenToken,
    Scope = "openid profile email our_user_id https://www.our-domain.com.au/groups",
    TokenValidationParameters = new TokenValidationParameters { NameClaimType = "name" },
    Notifications = new OpenIdConnectAuthenticationNotifications { ... }
    CookieManager = new SameSiteCookieManager(new SystemWebCookieManager()) // I've added this!!
});
```

What's not 100% clear to me is whether our flow will be affected by this cookie change. Auth0 docs say response_mode=form_post are a concern, but we aren't explicitly setting the ResponseMode in the config and our ResponseType = CodeIdTokenToken, which apparently has a default ResponseMode of Fragment.

I've followed the advice in the [asp.net blog post](https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/) that says to hookup the custom `ICookieManager`, which is passed the `SystemWebCookieManager` you mentioned. I've also added it to the `CookieAuthenticationOptions` provided to the `UseCookieAuthentication` call, but not sure if that's required.

If we don't need to do anything here, I'm happy not too. Otherwise like you mentioned, we can just force everyone to Owin 4.1 and netfx472 as they really need to be on it these days I guess.

benmccallum commented:
On that last note, I don't think we care if we are tied to System.Web in our Owin code for this app. The intention is to get this app onto dotnetcore soon, just trying to figure out when :P Perhaps we've already tied it to System.Web anyway by following the advice in the blog post.

benmccallum commented:
Is the "Cookie Monster" bug even still an issue in Owin 4 and 4.1?

AndersAbel commented:
> Is the "Cookie Monster" bug even still an issue in Owin 4 and 4.1?

Yes it is. The reason for the bug is deep within System.Web and cannot be easily changed.

If you have followed the advice and hooked up a SystemWebCookieManager you should remove the Kentor.OwinCookieSaver package from your solution. The SystemWebCookieManager (wrapped in your SameSiteCookieManager) already handles the cookie monster bug.

benmccallum commented:
Cool, that should sort it for us then. Will leave this PR on standby though in case anyone needs it. It would just needs those Owin dependencies updated if desired.

Tratcher commented:
> What's not 100% clear to me is whether our flow will be affected by this cookie change. Auth0 docs say response_mode=form_post are a concern, but we aren't explicitly setting the ResponseMode in the config and our ResponseType = CodeIdTokenToken, which apparently has a default ResponseMode of Fragment.

UseOpenIdConnectAuthentication doesn't support Fragment, it's form_post by default and is directly affected by the SameSite changes.

> // @Tratcher any thoughts on how to help people migrate off this package and to SystemWebCookieManager instead?

Have you considered not updating this package? Updating the readme and redirecting people to SystemWebCookieManager would make your life easier.

benmccallum commented:
Thanks for the clarity @Tratcher! This has been a fun one to sort out.

benmccallum commented:
FYI, one thing that got me briefly, that could be made clearer in the MS announcement/docs, is that on localhost, you're really gonna have to move to https, as the SameSite=None cookies need Secure and therefore https. In my snippet above, we were doing `UseCookieAuthentication` passing options with `CookieSecure = CookieSecureOption.SameAsRequest`, though potentially `<httpCookie requireSSL="true">` was overriding that, but at the end of the day on http://localhost it obv wasn't working. 

I also updated the `CheckSameSite` method to this just in case:

```csharp
private void CheckSameSite(IOwinContext context, CookieOptions options)
        {
            if (options.SameSite == SameSiteMode.None)
            {
                if (DisallowsSameSiteNone(context))
                {
                    options.SameSite = null;
                }
                else
                {
                    // This is a fail-safe, probably overkill. Essentially, here we know
                    // that the cookie with be SameSite=None, so it should also be Secure
                    // as some browsers will start blocking None and non-secure soon. 
                    options.Secure = true;
                }
            }
        }
```

AndersAbel commented:
> Have you considered not updating this package? Updating the readme and redirecting people to SystemWebCookieManager would make your life easier.

I definitely have! I thought that this package was a thing of the past. But GitHub shows 400 daily downloads so it's obiously still used. Now I'm considering adding a dummy 2.0.0 version to Nuget that only contains a readme that describes how to use the SystemWebCookieManager instead.
